File: application/services/training/training.service.ts
import { ApplicationService } from "../application.service";
import { CreateTrainingCommand } from "../../commands/fitness/trainings/CreateTrainingCommand"
import { UpdateTrainingCommand } from "../../commands/fitness/trainings/UpdateTrainingCommand"
import { DeleteTrainingCommand } from "../../commands/fitness/trainings/DeleteTrainingCommand"
import { CreateTrainingVideoCommand } from "../../commands/fitness/trainings/CreateTrainingVideoCommand"
import { UpdateTrainingVideoCommand } from "../../commands/fitness/trainings/UpdateTrainingVideoCommand"
import { Training } from "core/fitness/trainings/Training";
import { TrainingRepository } from "../../repositories/fitness/training.repository.interface";
import { DeleteTrainingVideoCommand } from "../../commands/fitness/trainings/DeleteTrainingVideoCommand";

type Commands =
    CreateTrainingCommand
    | UpdateTrainingCommand
    | DeleteTrainingCommand
    | CreateTrainingVideoCommand
    | UpdateTrainingVideoCommand
    | DeleteTrainingVideoCommand


export class TrainingService implements
    ApplicationService<CreateTrainingCommand, Training>,
    ApplicationService<UpdateTrainingCommand, Training | undefined>,
    ApplicationService<DeleteTrainingCommand, undefined>,
    ApplicationService<CreateTrainingVideoCommand, Training>,
    ApplicationService<UpdateTrainingVideoCommand, boolean | undefined>,
    ApplicationService<DeleteTrainingVideoCommand, true | undefined>
{

    constructor(private repository: TrainingRepository) { }

    handle(command: UpdateTrainingCommand): Promise<Training | undefined>
    handle(command: CreateTrainingCommand): Promise<Training>
    handle(command: DeleteTrainingCommand): Promise<undefined>
    handle(command: CreateTrainingVideoCommand): Promise<Training>
    handle(command: UpdateTrainingVideoCommand): Promise<Training | undefined>
    handle(command: DeleteTrainingVideoCommand): Promise<true | undefined>
    async handle(command: Commands) {
        if (command instanceof CreateTrainingCommand) {
            const training = Training.create({
                name: command.name,
                description: command.description,
                trainerId: command.trainerId,
                categories: command.categories
            })
            await this.repository.save(training)

            return training
        } else if (command instanceof UpdateTrainingCommand) {
            const training = await this.repository.find(command.id)
            if (training) {
                training.update(command.data)
                await this.repository.save(training)

                return training
            }
        } else if (command instanceof DeleteTrainingCommand) {
            await this.handle(new DeleteTrainingVideoCommand(command.id))

            const training = await this.repository.find(command.id)
            if (training) {
                await this.repository.delete(training)
                training.destroy()
            }
        } else if (command instanceof CreateTrainingVideoCommand) {
            const training = await this.repository.find(command.trainingId),
                videoBuffer = Buffer.from(command.video, 'base64')

            training!.setVideo(command.filename, videoBuffer)
            return await this.repository.addVideo(
                training!,
                videoBuffer
            )
        } else if (command instanceof UpdateTrainingVideoCommand) {
            const training = await this.repository.find(command.trainingId).then(t => t!)

            if (!command.video) {
                const videoBuffer = await this.repository.getVideo(training)
                if (!videoBuffer) return
                training.setVideo(command.filename, videoBuffer, command.ext)
                return await this.repository.addVideo(training)
            } else {
                const videoBuffer = Buffer.from(command.video, 'base64')
                training.setVideo(command.filename, videoBuffer, command.ext)
                return await this.repository.addVideo(training, videoBuffer)
            }

        } else if (command instanceof DeleteTrainingVideoCommand) {
            const training = await this.repository.find(command.trainingId)
            if (training) {
                await this.repository.deleteVideo(training)
                training.destroyVideo()
                return true
            }
        }
    }
}

File: fitness-dimension/application/trainings/training_service.go
package trainings

import (
	"fitness-dimension/application/trainings/commands"

	"fitness-dimension/core/trainings/training"
	"fitness-dimension/core/trainings/training/entities"
	valuesObjects "fitness-dimension/core/trainings/training/values-objects"
)

type TrainingService struct {
	Publisher  TrainingPublisher
	Repository TrainingRepository
}

func (s *TrainingService) Handle(c interface{}) (interface{}, error) {
	switch c.(type) {
	case commands.CreateTraining:
		command := c.(commands.CreateTraining)

		trainerID := valuesObjects.TrainerID{Value: command.TrainerID}
		name := valuesObjects.TrainingName{Value: command.Name}
		description := valuesObjects.TrainingDescription{Value: command.Description}
		categories := valuesObjects.TrainingTaxonomies{Values: command.Categories}

		t, _ := training.CreateTraining(categories, trainerID, name, description)
		for _, i := range t.GetEvents() {
			s.Publisher.Publish(i)
		}

		return t.ID.Value.String(), nil

	case commands.UpdateTraining:
		command := c.(commands.UpdateTraining)

		trainerID := valuesObjects.TrainerID{Value: command.TrainerID}
		name := valuesObjects.TrainingName{Value: command.Name}
		description := valuesObjects.TrainingDescription{Value: command.Description}
		categories := valuesObjects.TrainingTaxonomies{Values: command.Categories}

		t, err := s.Repository.Get(command.ID.String())
		if err != nil {
			return nil, err
		}

		t.Update(categories, trainerID, name, description)
		for _, i := range t.GetEvents() {
			err := s.Publisher.Publish(i)
			if err != nil {
				return nil, err
			}
		}

	case commands.DeleteTraining:
		command := c.(commands.DeleteTraining)

		t, err := s.Repository.Get(command.ID.String())
		if err != nil {
			return nil, err
		}

		t.Destroy()
		for _, i := range t.GetEvents() {
			err := s.Publisher.Publish(i)
			if err != nil {
				return nil, err
			}
		}

	case commands.CreateTrainingVideo:
		command := c.(commands.CreateTrainingVideo)

		filename := valuesObjects.TrainingVideoName{Value: command.Name}
		video := valuesObjects.TrainingVideoBuffer{Value: command.Video}
		ext := valuesObjects.TrainingVideoExt{Value: command.Ext}

		t, err := s.Repository.Get(command.TrainingID.String())
		if err != nil {
			return nil, err
		}

		t.SetVideo(filename, video, ext)

		for _, i := range t.GetEvents() {
			err := s.Publisher.Publish(i)
			if err != nil {
				return nil, err
			}
		}

	case commands.UpdateTrainingVideo:
		command := c.(commands.UpdateTrainingVideo)

		filename := valuesObjects.TrainingVideoName{Value: command.Name}
		video := valuesObjects.TrainingVideoBuffer{Value: command.Video}
		ext := valuesObjects.TrainingVideoExt{Value: command.Ext}

		t, err := s.Repository.Get(command.TrainingID.String())
		if err != nil {
			return nil, err
		}

		t.UpdateVideo(filename, video, ext)

		for _, i := range t.GetEvents() {
			err := s.Publisher.Publish(i)
			if err != nil {
				return nil, err
			}
		}

	case commands.DeleteTrainingVideo:
		command := c.(commands.DeleteTrainingVideo)

		t, err := s.Repository.Get(command.TrainingID.String())
		if err != nil {
			return nil, err
		}

		t.DestroyVideo()
		for _, i := range t.GetEvents() {
			err := s.Publisher.Publish(i)
			if err != nil {
				return nil, err
			}
		}
	}
	return nil, nil
}

func (s *TrainingService) Get(id string) (*training.Training, error) {
	return s.Repository.Get(id)
}

func (s *TrainingService) GetAll() ([]training.Training, error) {
	return s.Repository.GetAll()
}

func (s *TrainingService) GetByTrainer(id string) ([]training.Training, error) {
	return s.Repository.GetByTrainer(id)
}

func (s *TrainingService) GetVideo(id string) *entities.TrainingVideo {
	return s.Repository.GetVideo(id)
}
